src/coap_blockwise.c

```diff
@@ -22,13 +22,152 @@ struct blockwise_transfer
     uint8_t *block_buffer;
     union
     {
-        // TODO: add read_block_cb here
+        read_block_cb read_cb;
         write_block_cb write_cb;
     } callback;
     void *callback_arg;
 };
 
-// Function to initialize the blockwise_transfer structure
+/* Blockwise Uploads related functions */
+
+// Function to initialize the blockwise_transfer structure for uploads
+static void blockwise_upload_init(struct blockwise_transfer *ctx,
+                                  uint8_t *data_buf,
+                                  const char *path_prefix,
+                                  const char *path,
+                                  read_block_cb cb,
+                                  void *callback_arg)
+{
+    ctx->is_last = false;
+    ctx->path_prefix = path_prefix;
+    ctx->path = path;
+    ctx->retry_count = 0;
+    ctx->block_size = 0;
+    ctx->offset = 0;
+    ctx->block_buffer = data_buf;
+    ctx->callback_arg = callback_arg;
+    ctx->callback.read_cb = cb;
+}
+
+// Blockwise upload's internal callback function that the COAP client calls
+static void on_block_sent(struct golioth_client *client,
+                          const struct golioth_response *response,
+                          const char *path,
+                          void *arg)
+{
+    assert(arg);
+    struct blockwise_transfer *ctx = (struct blockwise_transfer *) arg;
+    ctx->status = response->status;
+}
+
+// Function to call the application's read block callback after a successful
+// blockwise upload
+static void call_read_block_callback(struct blockwise_transfer *ctx)
+{
+    if (ctx->callback.read_cb(ctx->offset,
+                              ctx->block_buffer,
+                              &ctx->block_size,
+                              &ctx->is_last,
+                              ctx->callback_arg)
+        != GOLIOTH_OK)
```

Comment:
I think this would be easier to read.
```suggestion
    int err = ctx->callback.read_cb(ctx->offset,
                                    ctx->block_buffer,
                                    &ctx->block_size,
                                    &ctx->is_last,
                                    ctx->callback_arg);
    if (err != GOLIOTH_OK)
```